plugins/user-authenticators/ldap/src/main/java/org/apache/cloudstack/api/command/LdapListUsersCmd.java

```diff
@@ -66,34 +121,113 @@ public LdapListUsersCmd(final LdapManager ldapManager, final QueryService queryS
+        // if we are filtering for local domain, only get users for the current domain
+        // if we are filtering for any domain, get recursive all users for the root domain
+        // if we are filtering for potential imports,
+        //    we are only looking for users in the linked domains/accounts,
+        //    which is only relevant if we ask ldap users for this domain.
+        //    So we are asking for all users in the current domain as well
+        // in case of no filter we should find all users in the current domain for annotation.
+        if (cloudstackUsers == null) {
+            ListResponse<UserResponse> cloudstackUsersresponse;
+            switch (getUserFilter()) {
+            case ANY_DOMAIN:
+                cloudstackUsersresponse = _queryService.searchForUsers(1l,true);
+                break;
+            case NO_FILTER:
+                cloudstackUsersresponse = _queryService.searchForUsers(this.domainId,true);
+                break;
+            case POTENTIAL_IMPORT:
+            case LOCAL_DOMAIN:
+                cloudstackUsersresponse = _queryService.searchForUsers(this.domainId,false);
+                break;
+            default:
+                throw new CloudRuntimeException("error in program login; we are not filtering but still querying users to filter???");
+            }
+            cloudstackUsers = cloudstackUsersresponse.getResponses();
+            if(s_logger.isTraceEnabled()) {
+                StringBuilder users = new StringBuilder();
+                for (UserResponse user : cloudstackUsers) {
+                    if (users.length()> 0) {
+                        users.append(", ");
+                    }
+                    users.append(user.getUsername());
+                }
+
+                s_logger.trace(String.format("checking against %d cloudstackusers: %s.", this.cloudstackUsers.size(), users.toString()));
+            }
+        }
+        return cloudstackUsers;
+    }
+
+    private List<LdapUserResponse> applyUserFilter(List<LdapUserResponse> ldapResponses) {
+        if(s_logger.isTraceEnabled()) {
+            s_logger.trace(String.format("applying filter: %s or %s.", this.getListTypeString(), this.getUserFilter()));
+        }
+        Method filterMethod = getFilterMethod();
+        List<LdapUserResponse> responseList = ldapResponses;
+        if (filterMethod != null) {
+            if(s_logger.isTraceEnabled()) {
+                s_logger.trace("applying filter: " + filterMethod.getName());
+            }
+            try {
+                responseList = (List<LdapUserResponse>)filterMethod.invoke(this, ldapResponses);
+            } catch (IllegalAccessException | InvocationTargetException e) {
+                throw new CloudRuntimeException("we're damned, the earth is screwed. we will now return to our maker",e);
+            }
+        }
+        return responseList;
+    }
+
+    Method getFilterMethod() {
```

Comment:
For someone not well-versed with reflections and this style of programming, can you explore a way to re-implement the logic without using runtime dependent logic. The user filter type is already defined as an enum, is there a way to simply define what method to run based on the enum; i.e. the logic of method could be defined in the enum itself on how it applies the filter for example https://stackoverflow.com/questions/18883646/java-enum-methods/18883775#18883775

Reply:
worth a quick shot, will look into it

Reply:
@rhtyd I did see too much tight coupling to address this in a short time. Can you live with the current state?